ffs/scheduler/scheduler.go

```diff
@@ -396,28 +459,29 @@ func (s *Scheduler) executeQueuedStorage(j ffs.StorageJob) {
 	cancelChan := make(chan struct{}, 1)
 	// Create chan to allow Job cancellation.
 	s.cancelLock.Lock()
-	s.cancelChans[j.ID] = cancelChan
+	s.jobsCancel[j.ID] = cancelChan
 	s.cancelLock.Unlock()
 	defer func() {
 		s.cancelLock.Lock()
-		delete(s.cancelChans, j.ID)
+		delete(s.jobsCancel, j.ID)
 		s.cancelLock.Unlock()
 	}()
 
 	ctx, cancel := context.WithCancel(context.WithValue(context.Background(), ffs.CtxKeyJid, j.ID))
 	defer cancel()
 	ctx = context.WithValue(ctx, ffs.CtxStorageCid, j.Cid)
+	ctx = context.WithValue(ctx, ffs.CtxAPIID, j.APIID)
 
 	var cancelLock sync.Mutex
 	var canceled bool
 	go func() {
 		// If the user called Cancel to cancel Job execution,
 		// we cancel the context to finish.
 		<-cancelChan
-		cancel()
 		cancelLock.Lock()
 		canceled = true
 		cancelLock.Unlock()
+		cancel()
```

Comment:
A fix here. We first should set `canceled = true` and then cancel the context. Doing otherwise might produce canceled logic to still read `canceled = true` if they run fast enough.